hack/generator/spec/defaulter-validator.md

```diff
@@ -0,0 +1,100 @@
+`controller-runtime` defines `admission.Defaulter` and `admission.Validator`. These interfaces only give you a single `Default` or `ValidateX` method, which means that all validation/defaulting needs to be done in that method. I think we're going to quickly run into situations where we want custom (handcrafted) validations or defaults for a particular resource and we're not going to want to teach the code generator about these.
+
+## Suggestion
+Structure our autogenerated webhooks such that they're a pipeline. Add the ability to "take control" of this process by implementing an interface.
+
+Implement one (or both) of these interfaces to take control of validation/defaulting by hand
+```
+type Validator interface {
+	CreateValidations() []func() error
+	UpdateValidations() []func(old runtime.Object) error
+	DeleteValidations() []func() error
+}
+
+type Defaulter interface {
+	Defaults() []func()
+}
+```
+
+Sample `Default`:
+```
+// +kubebuilder:webhook:path=/mutate-microsoft-storage-infra-azure-com-v1alpha1api20190401-storageaccount,mutating=true,sideEffects=None,matchPolicy=Exact,failurePolicy=fail,groups=microsoft.storage.infra.azure.com,resources=storageaccounts,verbs=create;update,versions=v1alpha1api20190401,name=default.v1alpha1api20190401.storageaccounts.microsoft.storage.infra.azure.com,admissionReviewVersions=v1beta1
+
+var _ admission.Defaulter = &StorageAccount{}
+
+// Default defaults the Azure name of the resource to the Kubernetes name
+func (storageAccount *StorageAccount) Default() {
+	defaults := storageAccount.defaults()
+	var temp runtime.Object = storageAccount
+	if runtimeDefaulter, ok := temp.(genruntime.Defaulter); ok {
+		defaults = runtimeDefaulter.Defaults()
+	}
+
+	for _, defaultFunc := range defaults {
+		defaultFunc()
+	}
+}
+
+func (storageAccount *StorageAccount) defaults() []func() {
+	return []func() {
+		storageAccount.defaultAzureName,
+	}
+}
+
+func (storageAccount *StorageAccount) defaultAzureName() {
+	if storageAccount.Spec.AzureName == "" {
+		storageAccount.Spec.AzureName = storageAccount.Name
+	}
+}
+```
+
+Sample `Validate`: 
+```
+// +kubebuilder:webhook:path=/validate-microsoft-storage-infra-azure-com-v1alpha1api20190401-storageaccount,mutating=false,sideEffects=None,matchPolicy=Exact,failurePolicy=fail,groups=microsoft.storage.infra.azure.com,resources=storageaccounts,verbs=create;update,versions=v1alpha1api20190401,name=validate.v1alpha1api20190401.storageaccounts.microsoft.storage.infra.azure.com,admissionReviewVersions=v1beta1
+
+var _ admission.Validator = &StorageAccount{}
+
+// ValidateCreate validates the creation of the resource
+func (storageAccount *StorageAccount) ValidateCreate() error {
+	validations := storageAccount.createValidations()
+	var temp runtime.Object = storageAccount
+	if runtimeValidator, ok := temp.(genruntime.Validator); ok {
+		validations = runtimeValidator.CreateValidations()
+	}
+
+	for _, validation := range validations {
+		err := validation()
+		if err != nil {
```

Comment:
It might be worth running all the validations and aggregating errors here.

Reply:
Updated